src/core/solvers/ddp.cpp

```diff
@@ -16,7 +16,7 @@ SolverDDP::SolverDDP(boost::shared_ptr<ShootingProblem> problem)
     : SolverAbstract(problem),
       reg_incfactor_(10.),
       reg_decfactor_(10.),
-      reg_min_(1e-9),
+      reg_min_(1e-18),
```

Comment:
This should be ZERO.

Reply:
A zero value is not efficient because it makes our algorithm to decrease the damping factor to a level that doesn't influence the direction.
Furthermore, it adds unnecessarily iterations to increase the damping factor if the algorithm is recovering for very low-still not useful-value.

Reply:
I believe this topic merits a proper analysis. But this is very hard and out of the scope of this PR.

In any case, thank you for the feedback!

Reply:
I'm a bit afraid of your answer. I don't think you realize what you are writing so far.

If you want to solve a strictly feasible LQR problem, you don't need any regularization term at ALL.
While with your value, you are shifting your problem and then the solution!

In addition, `1e-18` makes totally no sense when thinking in terms of double arithmetic. 

Please @cmastalli don't escape from other contributors' remarks with very unfounded answers !!! It is very painful in the end !!!

Reply:
@jcarpent I appreciate your feedback but this is not the right way of working.
Let me explain you why I believe this is not the right way of working.

This PR consists on adding the possibility of having two factors for the regularization.
Then, you asked two things which are out of that scope:
  1. Remove const reference for primitives
  2. Change a value that historical was there

Reviewing other parts of the code and asking to change then is not productive, and prone error as well.
It is not productive because the developer can't be focused on the PR itself!
For me, it is useful is you write an issue, and then I could allocate time for making a careful work.

However, for this case, I follow both suggestions.
You need to be aware that the point 2 is a delicate changed.
I have designed that value based on running our examples.
I encountered that having zero increases the number of iterations in a couple of problems.
As you might see, I don't have time to report each result for a requested that it is out of the scope.

Reply:
I don't escape for discussions that have a clear solution.
But to keep me productivity, I will do it if the discussion is going anywhere!

Reply:
Btw, the value is `1e-16`. I just forgot to push it.

Reply:
Let me know if this solves your point.

And again, note that we have a routine that decreases the damping value. There is no sense to decrease it more than `1e-16`

Reply:
1) I agree that we should start this value from zero. It doesn't make any sense to use such ridiculously low value (putting 1e-18 or 1e-16 is non-sense).

2) For the reg_max, I don't understand why you decided to increase it to 1e12? 1e9 is already quite high, and if problem cannot be solved with 1e9 as the regularization, there is something wrong with the problem. 
1e9 is already an arbitrary choice, and so is 1e12. Perhaps it would be worth setting the max reg factor as a value relative 
to the problem Quu? But that would be too complicated. 


Regarding the scope of this PR:
1) Regularization is meant to help an improperly formed shooting problem. However, currently we are helping every problem, regardless of whether it is needed or not. So this was a problem in the initial design, which we have a chance of fixing now. So thanks @cmastalli for taking care of that.

2) There is already a bunch of mixed stuff in this PR. So it feels normal that when perusing the files, you can spot issues with existing code. You are within your rights @cmastalli to defer suggestions to a future PR (while keeping track of the suggestions). But once you have accepted and implemented the suggestion, the discussion has to be complete and the implementation proper. Changing 1e-9 to 1e-16 is not the solution of Justin's suggestion.

Reply:
Hi @proyan,

I don't agree with the suggestions in this thread.
I would explain you why, but if this doesn't yet convinced you then we need to discuss it in private.

>     1. I agree that we should start this value from zero. It doesn't make any sense to use such ridiculously low value (putting 1e-18 or 1e-16 is non-sense).

No, it is not a good idea to define the minimum regularization (or to start the algorithm) with a zero value.
First, values lower than `1e-16` (i.e., the double resolution) do not really affect the direction or do not produce any numerical effect. I have also tested this well-known fact across our problems. Please believe me or try yourself.
Second, if you use define the minimum regulation value to zero, you are allowing our algorithm to decrease superfluously  the regularization value. Furthermore, it would take useless steps to recover from it, i.e., to increase bigger than `1e-16`.
I have also validated this statement across our problems. Indeed, this suggestion increases the number of iteration in some cases.

Note that we have implemented a Levenberg-Marquardt algorithm. It is common across LM implementation to have a minimum regularization values different to zero. Indeed, this is clearly stated in the original paper that proposed the LM strategy I copied for you the text below:
![image](https://user-images.githubusercontent.com/3601935/103558739-fe234d80-4eac-11eb-8b82-9137b2ae91be.png)
 PDF: https://www.google.com/url?sa=t&rct=j&q=&esrc=s&source=web&cd=&ved=2ahUKEwiM3OHei7ztAhVhRhUIHXu1DcEQFjAAegQIARAC&url=https%3A%2F%2Fepubs.stfc.ac.uk%2Fmanifestation%2F6683%2FAERE_R_6799.pdf&usg=AOvVaw19LGQ4NFbIsBebeSX7BgPg

---

>     2\. For the reg_max, I don't understand why you decided to increase it to 1e12? 1e9 is already quite high, and if problem cannot be solved with 1e9 as the regularization, there is something wrong with the problem.
>         1e9 is already an arbitrary choice, and so is 1e12. Perhaps it would be worth setting the max reg factor as a value relative
>         to the problem Quu? But that would be too complicated.
This number indeed needs to be studied more deeply. I just found that in some very hard problems is good to allow a bigger value.
Your proposition can be done for a specific problem (at the user level). With a bigger value, I am jusrt making the default configuration more robust.

I hope this follow-up explanation can satisfy you. Otherwise, you really need to make contra-argument with evidence. It is not fair to insist without any support of your statement. I am more than open to get feedback but with evidence!


**Note**
In near future, I will conduct another study to potentially improve a bit more the regularization. For instance, it might be useful to always start with Gauss-Newton direction as KNITRO performs. Note that we always start with the stored regularization value. I have been spending quite a lot of my time in these tiny details, please do not think that I am taking this lightly.

Reply:
I have reverted the values as discussed with @nmansard.
Note that the values were `1e-9` and `1e9` for the `regMin` and `regMax`, respectively.